core/controllers/reader_test.py

```diff
@@ -362,3 +362,112 @@ def test_ratings_by_different_users(self):
+            category='This is just a spam category',
+            objective='Test a spam exploration.')
+        self.can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS', True)
+        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+        exp_services._save_exploration(self.editor_id, exploration, '', [])  # pylint: disable=protected-access
+        rights_manager.publish_exploration(self.editor_id, self.EXP_ID)
+        self.logout()
+
+    def test_that_emails_are_sent(self):
+        # Login and flag exploration.
+        self.login(self.NEW_USER_EMAIL)
+
+        response = self.testapp.get('/explore/%s' % self.EXP_ID)
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        # Create report for exploration.
+        self.post_json(
+            '%s/%s' % (feconf.FLAG_EXPLORATION_URL_PREFIX, self.EXP_ID), {
+                'report_text': self.REPORT_TEXT,
+            }, csrf_token)
+
+        self.logout()
+
+        expected_email_html_body = (
+            'Hello Moderator,<br>'
+            'newuser has submitted a new report on the exploration '
+            'Welcome to Oppia!'
+            ' on the grounds of: <br>'
+            'AD .<br>'
+            'You can modify the exploration by clicking '
+            '<a href="https://www.oppia.org/create/0">'
+            '"here"</a>.<br>'
+            '<br>'
+            'Thanks!<br>'
+            '- The Oppia Team<br>'
+            '<br>'
+            'You can change your email preferences via the '
+            '<a href="https://www.example.com">Preferences</a> page.')
+
+        expected_email_text_body = (
+            'Hello Moderator,\n'
+            'newuser has submitted a new report on the exploration '
+            'Welcome to Oppia!'
+            ' on the grounds of: \n'
+            'AD .\n'
+            'You can modify the exploration by clicking "here".\n'
+            '\n'
+            'Thanks!\n'
+            '- The Oppia Team\n'
+            '\n'
+            'You can change your email preferences via the Preferences page.')
+
+        with self.can_send_emails_ctx:
+            self.process_and_flush_pending_tasks()
+
+            messages = self.mail_stub.get_sent_messages(to=self.MODERATOR_EMAIL)
+            self.assertEqual(len(messages), 1)
+            self.assertEqual(
+                messages[0].html.decode(),
+                expected_email_html_body)
+            self.assertEqual(
+                messages[0].body.decode(),
+                expected_email_text_body)
+
+    def test_that_emails_are_not_sent(self):
+        """
+        response = self.testapp.get('/explore/%s' % self.EXP_ID)
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        # Create report for exploration.
+        self.post_json(
+            '%s/%s' % (feconf.FLAG_EXPLORATION_URL_PREFIX, self.EXP_ID), {
+                'report_text': self.REPORT_TEXT,
+            }, csrf_token)
```

Comment:
I don't know, I'm afraid, but here are some things I would try:

(1) If you raise an Exception with the csrf_token just above this, is it None?

(2) Try replicating the conditions of this test manually, and look at the terminal to see which URL routes are being hit. Do you see any 302s?